File: Julia_Code/OLG_Aiyagari/Main_Solve_OLG_Model.jl
# Computing Longitudinal Moments for Heterogeneous Agent Models
# Sergio Ocampo, Baxter Robinson, and Emmanuel Murray Leclair
# April 2022
# Aiyagari economy: 
#       1. Finitely lived agents and overlapping generations
#       2. Inelastic labor supply
# This scripts computes longitudinal moments for the model
# 
# Solve:   V(h,ϵ,a) = max{ ((1+r)a+wϵ̄f(h,ϵ)-a')^(1-γ)/(1-γ) + beta*E[ s(h)*V(h+1,ϵ',a') + (1-s(h))*ν(a')|ϵ] }
#           log(ϵ') = ρ_ϵ*log(ϵ) + η_ϵ; η_ϵ~N(0,σ_ϵ); 
# The constant ϵ̄ guarantees that E[f(h,ϵ)]=1 and so aggregate labor L=E[ϵ]=1

## Change to your home directory 
# Sergio's Computer 
#    cd()
#    cd("./Dropbox/Research/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
# Emmanuel's Computer
    # cd()
    # cd("C:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/") # Laptop
    # cd("D:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/") # Desktop
    # cd("C:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
# Baxter's Computer
     cd("D:/Dropbox/Files/Economics-Research/Project-09_SIM/Code/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
# Compute Canada Server
    # cd("/scratch/robin370/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")

## Make auxiliary directores
    Fig_Folder  = "Figures" ; mkpath(Fig_Folder)  ;
    File_Folder = "Files"   ; mkpath(File_Folder) ;
    Hist_Folder = "Files/Histogram"  ; mkpath(Hist_Folder) ;
    MC_Folder   = "Files/MonteCarlo" ; mkpath(MC_Folder)   ;


# Load packages
using SparseArrays
using Plots
using Interpolations # Pkg.add("Interpolations") # https://github.com/JuliaMath/Interpolations.jl
using Dierckx # Pkg.add("Dierckx") # https://github.com/kbarbary/Dierckx.jl
using ForwardDiff # Pkg.add("ForwardDiff") # https://github.com/JuliaDiff/ForwardDiff.jl
using Optim # Pkg.add("Optim") # https://julianlsolvers.github.io/Optim.jl/stable/
using Optim: converged, maximum, maximizer, minimizer, iterations
using Roots # Pkg.add("Roots") # https://github.com/JuliaMath/Roots.jl
using Parameters # Pkg.add("Parameters") # https://github.com/mauro3/Parameters.jl
using Kronecker # Pkg.add("Kronecker") # https://michielstock.github.io/Kronecker.jl
using DelimitedFiles
using Printf
using StatsBase

# Load functions in VFI Toolbox
include("../VFI_Toolbox.jl")

println(" ")
println("------------------------")
println("Aiyagari in Julia")
println("PWD: ",pwd())
println("This code uses Plots, Interpolations, Dierckx, ForwardDiff, Optim, Roots, Parameters, ScaledInterpolation")
println("Solve OLG Aiyagari model with EGM and the histogram method")
println("------------------------")
println(" ")

#-----------------------------------------------------------
#-----------------------------------------------------------
# Load function with population and survival probabilities
include("Setup_Demographics.jl")

#-----------------------------------------------------------
#-----------------------------------------------------------
# Parameters and Model Structure
    # Generate structure for parameters using Parameters module
    # We can set default values for our parameters
    @with_kw struct Par
        # Model Parameters
        β::Float64   = 0.94   ; # Discount factor
        σ::Float64   = 2.0    ; # Relative risk aversion (utility) parameter
        ρ_ϵ::Float64 = 0.963  ; # Persistence of labor efficiency process
        σ_ϵ::Float64 = 0.162  ; # Standard deviation of labor efficiency innovation
        # Model prices (partial equilibrium)
        r::Float64 = 0.0379   ; # Target wealth weighted 3.79% average real return on net-worth (Fagereng et al. 2020)
        w::Float64 = 53.624   ; # U.S. (2019) - tens of thousands $
        # Borrowing constraint
        a_min::Float64 = 1E-4 ; # Borrowing constraint
        # Histogram iteration parameters
        Hist_max_iter       = 1000  ; # Maximum number of iterations
        Hist_tol            = 1E-6  ; # Tolerance for distance
        Hist_η              = 0.00  ; # Dampen factor
        # Minimum consumption for numerical optimization
        c_min::Float64      = 1E-16 ; 
        # Life cycle parameters 
        Max_Age             = 81    ; # Corresponds to 100 years old 
        Surv_Pr             = Survival_Probabilities_Bell_Miller(Max_Age) ; 
        Age_Π               = Age_Transition(Max_Age,Surv_Pr)             ;
        Age_PDF             = Age_Distribution(Age_Π)                     ;
    end

# Allocate paramters to object p for future calling
p = Par();

# Generate structure of model objects
    # Model 1 : Aiyagari economy with infinitely lived agents and constant rate of returns
    @with_kw struct Model
        # Parameters
        p::Par = Par() # Model parameters in their own structure
        # Assets Grid
        a_max::Float64  = 10000                      # Max node of a_grid
        θ_a::Float64    = 3.5                        # Curvature of a_grid
        θ_a_f::Float64  = 3.5                        # Curvature of a_grid_fine
        n_a::Int64      = 250                        # Size of a_grid
        n_a_fine::Int64 = 500                        # Size of fine grid for interpolation and distribution
        a_grid          = Make_Grid(n_a     ,θ_a  ,p.a_min,a_max,"Poly")  # a_grid for model solution
        a_grid_fine     = Make_Grid(n_a_fine,θ_a_f,p.a_min,a_max,"Poly")  # Fine grid for interpolation
        # Labor productivity process - Transitory 
        n_ϵ       = 15                                 # Size of ϵ_grid
        MP_ϵ      = Rouwenhorst95(p.ρ_ϵ,p.σ_ϵ,n_ϵ)     # Markov Process for ϵ
        ϵ_ref     = 1.038479216975849/sum(exp.(MP_ϵ.grid).*MP_ϵ.PDF)   # Reference level for labor efficiency 
        ϵ_grid    = ϵ_ref*exp.(MP_ϵ.grid)              # Grid in levels
        # Labor productivity process - Life Cycle 
        age_vec   = collect(1:p.Max_Age)
        log_ξ_grid= (60*(age_vec.-1).-(age_vec.-1).^2)./1800  # Process peaks at age 50, and by age 80 gives the same income as when newborn
        ξ_ref     = 1/sum(exp.(log_ξ_grid).*p.Age_PDF)   # Reference level for labor efficiency 
        ξ_grid    = ξ_ref*exp.(log_ξ_grid)             # Grid in levels
        # State matrices
        a_mat     = repeat(a_grid,1,n_ϵ,p.Max_Age)
        ϵ_mat     = repeat(ϵ_grid',n_a,1,p.Max_Age)
        ξ_mat     = repeat(reshape(ξ_grid,(1,1,p.Max_Age)),n_a,n_ϵ,1)
        a_mat_fine= repeat(a_grid_fine,1,n_ϵ,p.Max_Age)
        ϵ_mat_fine= repeat(ϵ_grid',n_a_fine,1,p.Max_Age)
        ξ_mat_fine= repeat(reshape(ξ_grid,(1,1,p.Max_Age)),n_a_fine,n_ϵ,1)
        a_mat_aϵ  = repeat(a_grid,1,n_ϵ)
        # Labor income matrices 
        y_mat     = p.w*ϵ_mat.*ξ_mat
        y_mat_fine= p.w*ϵ_mat_fine.*ξ_mat_fine
        # Value and policy functions
        V         = Array{Float64}(undef,n_a,n_ϵ,p.Max_Age)       # Value Function
        G_ap      = Array{Float64}(undef,n_a,n_ϵ,p.Max_Age)       # Policy Function for capital
        G_c       = Array{Float64}(undef,n_a,n_ϵ,p.Max_Age)       # Policy Function
        V_fine    = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age)  # Value Function on fine grid
        G_ap_fine = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age)  # Policy Function on fine grid
        G_c_fine  = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age)  # Policy Function on fine grid
        # Distribution
        for i_h=1;p.Max_Age
        for i_ϵ=1:n_ϵ
        Γ[:,i_ϵ,i_h] .= 1/(n_a_fine)*MP_ϵ.PDF[i_ϵ]*p.Age_PDF[i_h] # Distribution (initiliazed to uniform)
        end
        end
        H_ind     = Array{Int64}(undef,n_a_fine,n_ϵ,p.Max_Age)              # Index for discretization of savings choice 
        # H_ω       = Array{Int64}(undef,n_a_fine,n_ϵ,p.Max_Age)                    # Probability of transition lo low index in discretization
        H_ω_lo_s   = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age,n_ϵ)
        H_ω_hi_s   = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age,n_ϵ)
        H_ω_lo_d   = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age)
        H_ω_hi_d   = Array{Float64}(undef,n_a_fine,n_ϵ,p.Max_Age)
        # Misc
        read_flag = false # Boolean for reading results from file 
    end

M = Model();

# Load functions in Functions_ModelSolution (solve the model and find stationary distribution)
include("Functions_ModelSolution.jl")


# Execute model solution 
println("\n===============================================\n Solving Aiyagari with EGM-Histogram(loop)")
    
    @time M_Aiyagari = Aiyagari_Equilibrium(Model(read_flag=true));

println("===============================================\n")


# # Get stats and graphs for the solution of the model 
# include("PrintStats_MakeGraphs.jl")


# # Get moments from histogram method
# include("CalculateMoments_Histogram.jl")


# # Get moments from simulation
# include("CalculateMoments_MonteCarlo.jl")

# Add Simulation Functions
include("Functions_MonteCarlo.jl")


# Run Draft Moments for Graphs and Tables 
include("Draft_Results.jl")


println("\n===============================================\n\n    End of Script \n\n===============================================")

File: Matlab_Code/OLG_Aiyagari/Main_Solve_OLG_Model.m
% % Computing Longitudinal Moments for Heterogeneous Agent Models
% Sergio Ocampo, Baxter Robinson, Emmanuel Murray Leclair and Javier Martinez
% April 2022
% Aiyagari economy: 
%       1. Infinitely lived agents
%       2. Inelastic labor supply
%       3. Stochastic rate of returns
% This scripts computes longitudinal moments for the model
% 
% Solve:   V(ζ,ϵ,a) = max{ ((1+r(ζ))a+wϵ̄ϵ-a')^(1-σ)/(1-σ) +beta*E[V(ζ',ϵ',a')|ζ,ϵ] }
%           log(ϵ') = ρ_ϵ*log(ϵ) + η_ϵ; η_ϵ~N(0,σ_ϵ);
%           r(ζ)    = exp(ζ)r⋆    
%           log(ζ') = ρ_ζ*log(ζ) + η_ζ; η_ζ~N(0,σ_ζ); 
% The constant ϵ̄ guarantees that E[ϵ]=1 and so aggregate labor L=E[ϵ]=1

%% Change to your home directory 
% Sergio's Computer 
  % cd()
   %cd("./Dropbox/Research/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
% Emmanuel's Computer
    % cd()
    % cd("C:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/") # Laptop
    % cd("D:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/") # Desktop
    % cd("C:/Users/Emmanuel/Dropbox/RA_Sergio/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
% Baxter's Computer
    %  cd("D:/Dropbox/Files/Economics-Research/Project-09_SIM/Code/Histogram_Iteration/Julia_Code/OLG_Aiyagari/")
% Javier's Computer
     %Home
     cd '/Users/cyberdim/Dropbox/WESTERN_ECONOMICS/RA_Baxter/Matlab_Hist_Iter/OLG'
     % Laptop
     %cd '/Volumes/EHDD1/Dropbox/WESTERN_ECONOMICS/RA_Baxter/Matlab_Hist_Iter/OLG'
% Compute Canada Server
%    cd '/scratch/robin370/Histogram_Iteration/Julia_Code/OLG_Aiyagari/ '
clc
clear


% Make auxiliary directories
mkdir Fig_Folder
mkdir File_Folder
mkdir Hist_Folder
mkdir MC_Folder


disp(" ")
disp("------------------------")
disp("Aiyagari in Matlab")
disp("PWD: ")
disp(pwd)
disp("Solve Aiyagari model with EGM and the histogram method")
disp("------------------------")
disp(" ")

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Parameters

%beta: 0.9400
%sigma: 2
%rho_eps: 0.9630
%sigma_eps: 0.1620
%r: 0.0379
%w: 53.6240
%a_min: 1.0000e-03
%Hist_max_iter: 1000
%Hist_tol: 1.0000e-06
%Hist_eta: 0
%c_min: 1.0000e-16
%Max_Age: 81
%Surv_Pr: [81×1 double]
%Age_Pi: [81×81 double]
%Age_PDF: [81×1 double]

par = Functions_ModelSolution.Parameters(0.94, 2.0, 0.963,0.162,...
    0.0379, 53.624,1e-3,1000,1e-6,0.00, 1e-16,81);
p=par;

%par = struct('beta',0.94, 'sigma',2.0, 'rho_eps', 0.963, 'sigma_eps',0.162,...
%    'r',0.0379, 'w',53.624,...
%    'a_min',1e-3,...
%'Hist_max_iter',1000,'Hist_tol',1e-6,'Hist_eta',0.00,...
%    'c_min',1e-16, ...
%    'Max_Age',81);
%par.Surv_Pr = Setup_Demographics.Survival_Probabilities_Bell_Miller(par.Max_Age);
%par.Age_Pi = Setup_Demographics.Age_Transition(par.Max_Age, par.Surv_Pr);
%par.Age_PDF = Setup_Demographics.Age_Distribution(par.Age_Pi);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Model

%par: [1×1 struct]
%a_max: 10000
%theta_a: 3.5000
%theta_a_f: 3.5000
%n_a: 250
%n_a_fine: 500
%n_eps: 11
%read_flag: 0

Model = Functions_ModelSolution.ModelStructure(par,10000,3.5,3.5,250, 500,11,0 );
M= Model;
%Model = struct('p',par, ...
%    'a_max', 10000, 'theta_a', 3.5, 'theta_a_f', 3.5, 'n_a',250, 'n_a_fine',500);
%Model.a_grid = VFI_Toolbox.Make_Grid(Model.n_a, Model.theta_a, par.a_min, Model.a_max,"Poly");
%Model.a_grid_fine = VFI_Toolbox.Make_Grid(Model.n_a_fine, Model.theta_a_f, par.a_min, Model.a_max,"Poly");

% Labor productivity process
%Model.n_eps = 11; % Size of ϵ_grid
%Model.MP_eps = struct('N',{}, 'grid',{}, 'Pi',{}, 'PDF',{},'CDF',{});
%[Model.MP_eps(1).N,Model.MP_eps(1).grid,Model.MP_eps(1).Pi,Model.MP_eps(1).PDF,Model.MP_eps(1).CDF] = VFI_Toolbox.Rouwenhorst95(par.rho_eps,par.sigma_eps,Model.n_eps); % Markov Process for epsilon
%Model.eps_ref = 1.038479216975849/dot(exp(Model.MP_eps(1).grid),Model.MP_eps(1).PDF); % Reference level for labor efficiency 
%Model.eps_grid = Model.eps_ref.*exp(Model.MP_eps(1).grid); % Grid in levels

% Labor productivity process - Life Cycle 
%Model.age_vec = transpose(1:1:par.Max_Age) ;
%Model.log_xi_grid = (60*(Model.age_vec-1) - ((Model.age_vec-1).^2))/1800 ; % Process peaks at age 50, and by age 80 gives the same income as when newborn
%Model.xi_ref = 1/sum(exp(Model.log_xi_grid).*par.Age_PDF) ; % Reference level for labor efficiency 
%Model.xi_grid = Model.xi_ref*exp(Model.log_xi_grid) ; %Grid in levels

% State matrices
%Model.a_mat = repmat(transpose(Model.a_grid),[1,Model.n_eps,par.Max_Age]);
%Model.eps_mat = repmat(Model.eps_grid,[Model.n_a,1,par.Max_Age]);
%Model.xi_mat = repmat(reshape(Model.xi_grid,[1,1,par.Max_Age]),Model.n_a, Model.n_eps,1);
%Model.a_mat_fine =  repmat(transpose(Model.a_grid_fine),[1,Model.n_eps,par.Max_Age]);
%Model.eps_mat_fine = repmat(Model.eps_grid,[Model.n_a_fine,1,par.Max_Age]);
%Model.xi_mat_fine = repmat(reshape(Model.xi_grid,[1,1,par.Max_Age]),Model.n_a_fine, Model.n_eps,1);
%Model.a_mat_aeps = repmat(transpose(Model.a_grid),1,Model.n_eps);

% Labor income matrices
%Model.y_mat = par.w*Model.eps_mat.*Model.xi_mat ; 
%Model.y_mat_fine = par.w*Model.eps_mat_fine.*Model.xi_mat_fine ; 

% Value and Policy Functions
%Model.V = zeros([Model.n_a, Model.n_eps, par.Max_Age]); % Value function
%Model.G_ap = zeros([Model.n_a, Model.n_eps, par.Max_Age]); % Policy Function for Capital/Assets
%Model.G_c = zeros([Model.n_a, Model.n_eps, par.Max_Age]); % Policy Function for Consumption
%Model.V_fine = zeros([Model.n_a_fine, Model.n_eps, par.Max_Age]); % Value function
%Model.G_ap_fine = zeros([Model.n_a_fine, Model.n_eps, par.Max_Age]); % Policy Function for Capital/Assets
%Model.G_c_fine = zeros([Model.n_a_fine, Model.n_eps, par.Max_Age]); % Policy Function for Consumption

% Distribution Guess
%Model.n_cut_fine = double(VFI_Toolbox.Grid_Inv(1000,Model.n_a_fine,Model.theta_a_f, par.a_min, Model.a_max,"Poly")); %  Index just below 1000
%Model.Gamma_a_guess = (1/((Model.n_cut_fine))).*([ones([Model.n_cut_fine, Model.n_eps, par.Max_Age]) ; zeros([Model.n_a_fine - Model.n_cut_fine, Model.n_eps, par.Max_Age])]);
%Model.Gamma = Model.Gamma_a_guess.*repmat(Model.MP_eps(1).PDF, Model.n_a_fine, 1, par.Max_Age).*repmat(reshape(par.Age_PDF,[1,1,par.Max_Age]),Model.n_a_fine, Model.n_eps,1);

% Matrices for discretization of policy functions
%Model.H_ind = zeros([Model.n_a_fine, Model.n_eps, par.Max_Age]); % Index for discretization of savings choice  
%Model.H_omega_lo_s = zeros(Model.n_a_fine, Model.n_eps, par.Max_Age, Model.n_eps); 
%Model.H_omega_hi_s = zeros(Model.n_a_fine, Model.n_eps, par.Max_Age, Model.n_eps);
%Model.H_omega_lo_d = zeros(Model.n_a_fine, Model.n_eps, par.Max_Age, Model.n_eps);
%Model.H_omega_hi_d = zeros(Model.n_a_fine, Model.n_eps, par.Max_Age, Model.n_eps);
% Misc 
%Model.read_flag = false; % Boolean for reading results from file. 

% End of model


% Execute model solution
disp('===============================================')
disp('Solving Aiyagari with EGM-Histogram(loop)')
%format longG
%
%
% 
%

tStart = tic;
M_Aiyagari = Functions_ModelSolution.Aiyagari_Equilibrium(Model);
tEnd = toc(tStart);

% Get stats and graphs for the solution of the model 
run('CalculateMoments_Solution.m')

disp('===============================================')




% Run Draft Moments for Graphs and Tables 
% include("Draft_Results.jl")



disp('=============================================== End of Script ===============================================')








